streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -153,7 +153,7 @@ public String apply(K key, V value) {
                 new KeyValueMapper<K, V, KeyValue<K1, V>>() {
                     @Override
                     public KeyValue<K1, V> apply(K key, V value) {
-                        return new KeyValue<>(mapper.apply(key, value), value);
+                        return (KeyValue<K1, V>) new KeyValue<>(mapper.apply(key, value), value);
```

Comment:
I don't think we need this cast. Do we?

Reply:
We do, becuase then the inferred type parameteres is correct. The cast is used to correct it.

Reply:
Cannot follow here. Do you get any compiler warning that the type cannot be inferred?

Using `new KeyValue<>` without type specification allow the compiler to infer the type automatically -- it's a short version of `new KeyValue<K1, V>` -- thus, even if the compiler would not be able to infer the type (but I am sure it does -- I don't see any warning) one should specify the type directly on the constructor call instead of adding a cast.

Does this make sense?